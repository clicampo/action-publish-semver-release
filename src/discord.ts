import fetch from 'node-fetch'
import * as core from '@actions/core'

export const notifyDiscordChannel = async(webhookUrl: string, options: {
    projectName: string
    projectUrl: string
    productionActionUrl: string
    nextVersion: string
    changelog: string
    isReleaseCandidate: boolean
}) => {
    core.startGroup('Notifying Discord channel')
    const version = options.nextVersion + (options.isReleaseCandidate ? '-rc' : '')
    const payload = {
        username: '',
        avatar_url: '',
        content: '**ðŸ”¥ tÃ¡ saindo do forninho versÃ£o nova**',
        embeds: [] as any[],
        components: [] as any[],
    }
    payload.embeds.push({
        title: options.projectName,
        color: 16711756,
        description: options.changelog
            // replace headings with bold text
            .replace(/#+ ([^\n]+)/g, '*$1*')
            // replace links with slack link syntax
            .replace(/\[([^\]]+)\]\(([^\)]+)\)/g, '<$2|$1>')
            // replace - with â†’
            .replace(/\- /g, 'â†’')
            // replace ** with *
            .replace(/\*\*([^\*]+)\*\*/g, '*$1*'),
        timestamp: '',
        author: {
            name: `â†³ ${version}`,
            url: options.projectUrl,
        },
        image: {},
        thumbnail: {},
        footer: {},
        fields: [],
    })
    payload.components.push({
        type: 1,
        components: [
            {
                type: 2,
                style: 5,
                label: 'Publicar em produÃ§Ã£o',
                url: options.productionActionUrl,
            },
        ],
    })

    core.info(`Sending payload to Discord\n${JSON.stringify(payload, null, 4)}`)
    const response = await fetch(webhookUrl, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    })
    core.endGroup()
    return response
}
