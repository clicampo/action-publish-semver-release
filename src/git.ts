import { getExecOutput } from '@actions/exec'
import * as core from '@actions/core'

export const getLastGitTag = async(considerReleaseCandidates: boolean): Promise<string | null> => {
    try {
        core.startGroup('Getting last git tag')
        const { stdout: gitTagList, exitCode } = await getExecOutput(
            'git for-each-ref --sort=creatordate --format "%(refname)" refs/tags',
        )
        if (exitCode !== 0)
            throw Error
        console.log(`${gitTagList.split('\n').filter(ref => considerReleaseCandidates ? true : !ref.includes('-rc')).length} tags found`)
        const lastGitTag = gitTagList
            .split('\n')
            .filter(ref => considerReleaseCandidates ? true : !ref.includes('-rc'))
            .reverse()[0]
            .split('/')
            .pop()
        if (lastGitTag === undefined || lastGitTag === '') {
            core.info('No git tag found.')
            throw Error
        }
        core.info(`Last git tag: ${lastGitTag}`)
        core.endGroup()
        return lastGitTag
    }
    catch (e) {
        core.error('Could not get last git tag')
        return null
    }
}

export const getLastCommitMessage = async(): Promise<string | null> => {
    try {
        core.startGroup('Getting last commit message')
        const { stdout: lastCommitMessage, exitCode } = await getExecOutput(
            'git log -1 --pretty=%B --no-merges',
        )
        if (exitCode !== 0)
            throw Error
        core.endGroup()
        return lastCommitMessage
    }
    catch (e) {
        core.error('Could not get last commit message')
        return null
    }
}

const setGitCommiter = async(): Promise<void> => {
    try {
        const name = core.getInput('git-committer-name')
        const email = core.getInput('git-committer-email')
        if (name === '' || email === '')
            throw new Error('Git committer name and email are required')

        core.startGroup('Setting git commiter identity')
        const { exitCode: exitCodeName } = await getExecOutput(
            `git config --global user.name "${name}"`,
            [],
            { silent: true },
        )
        if (exitCodeName !== 0)
            throw new Error('Could not set git commiter name')

        const { exitCode: exitCodeEmail } = await getExecOutput(
            `git config --global user.email "${email}"`,
            [],
            { silent: true },
        )
        if (exitCodeEmail !== 0)
            throw new Error('Could not set git commiter email')
        core.info('Git commiter identity set.')
        core.endGroup()
    }
    catch (e: any) {
        core.error(`Could not set git commiter identity\n${e.message}`)
    }
}

export const tagCommit = async(nextVersion: string, isReleaseCandidate: boolean): Promise<void | null> => {
    try {
        await setGitCommiter()
        core.startGroup(`Tagging ${isReleaseCandidate ? 'release candidate' : 'version'} ${nextVersion}`)
        const version = `${nextVersion}${isReleaseCandidate ? '-rc' : ''}`
        const { exitCode: tagExitCode } = await getExecOutput(
            `git tag -a ${version} -m "Release ${nextVersion}"`,
        )
        if (tagExitCode !== 0)
            throw Error

        const { exitCode: pushExitCode } = await getExecOutput(
            'git push --tags',
        )
        if (pushExitCode !== 0)
            throw Error
        core.info(`Git tag ${version} pushed.`)
        core.endGroup()
    }
    catch (e) {
        core.error('Could not tag commit')
        return null
    }
}

export const deleteTag = async(tag: string): Promise<void | null> => {
    try {
        core.startGroup('Deleting tag')
        const { exitCode: deleteTagExitCode } = await getExecOutput(
            `git tag -d ${tag}`,
        )
        if (deleteTagExitCode !== 0)
            throw Error

        const { exitCode: pushExitCode } = await getExecOutput(
            `git push --delete origin ${tag}`,
        )
        if (pushExitCode !== 0)
            throw Error
        core.info(`Git tag ${tag} deleted.`)
        core.endGroup()
    }
    catch (e) {
        core.error('Could not delete tag')
        return null
    }
}
