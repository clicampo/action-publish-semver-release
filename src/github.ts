import { getInput } from '@actions/core'
import { getOctokit } from '@actions/github'
import * as core from '@actions/core'
import type { Context } from '@actions/github/lib/context'

export const createGithubRelease = async(context: Context, nextVersion: string, body: string) => {
    core.startGroup('Creating GitHub release')
    const githubToken = getInput('github-token') || process.env.GH_TOKEN
    if (githubToken === '' || githubToken === undefined)
        throw new Error('GitHub token is required')

    const client = getOctokit(githubToken).rest

    try {
        const {
            data: {
                url: releaseUrl,
            },
        } = await client.repos.createRelease(
            {
                repo: context.repo.owner,
                owner: context.repo.repo,
                tag_name: nextVersion,
                name: nextVersion,
                body,
                prerelease: true,
            },
        )
        core.info(`Created release at ${releaseUrl}`)
        core.endGroup()
        return releaseUrl
    }
    catch (e: any) {
        core.info('Could not create GitHub release')
        console.log(e)
        core.error(e.message)
        core.setFailed(e.message)
    }
}
