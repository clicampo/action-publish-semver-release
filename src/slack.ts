import fetch from 'node-fetch'

export const notifySlackChannel = async(webhookUrl: string, options: {
    projectName: string
    projectUrl: string
    productionActionUrl: string
    nextVersion: string
    changelog: string
    isReleaseCandidate: boolean
}) => {
    const version = options.nextVersion + (options.isReleaseCandidate ? '-rc' : '')
    const summaryBlock = {
        type: 'section',
        text: {
            type: 'mrkdwn',
            text: `The project *<${options.projectUrl}|${options.projectName}>* has just released the version *${version}*!`,
        },
    }
    const changelogBlock = {
        type: 'section',
        text: {
            type: 'mrkdwn',
            text: options.changelog
                // replace headings with bold text
                .replace(/#+ ([^\n]+)/g, '*$1*')
                // replace links with slack link syntax
                .replace(/\[([^\]]+)\]\(([^\)]+)\)/g, '<$2|$1>')
                // replace - with →
                .replace(/\-/g, '→'),
        },
    }
    const payload = {
        blocks: [
            {
                type: 'header',
                text: {
                    type: 'plain_text',
                    text: ':rocket: New release!',
                    emoji: true,
                },
            },
            summaryBlock,
            changelogBlock,
            options.isReleaseCandidate && options.productionActionUrl
                ? {
                    type: 'section',
                    text: {
                        type: 'mrkdwn',
                        text: '_⚠️ This is a release candidate_',
                    },
                    accessory: {
                        type: 'button',
                        text: {
                            type: 'plain_text',
                            text: 'Deploy to production',
                            emoji: true,
                        },
                        value: 'n/a',
                        url: options.productionActionUrl,
                        action_id: 'n/a',
                    },
                }
                : undefined,
        ].filter(Boolean),
    }
    const response = await fetch(webhookUrl, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
    })
    return response
}
