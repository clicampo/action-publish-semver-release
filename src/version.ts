import * as core from '@actions/core'
export type ReleaseType = 'patch' | 'minor' | 'major'

export const getReleaseTypeFromCommitMessage = (commitMessage: string): ReleaseType | null => {
    if (/feat(\([^/)]+\))?!/.test(commitMessage))
        return 'major'
    if (/feat/.test(commitMessage))
        return 'minor'
    if (/fix/.test(commitMessage))
        return 'patch'
    return null
}

export const getNextVersion = (currentVersion: string, releaseType: ReleaseType): string => {
    // verify that the current version is valid semver
    if (currentVersion.match(/^\d+\.\d+\.\d+(-[\w\d]+)?$/) === null) {
        core.error(`Invalid current version: ${currentVersion}`)
        throw Error
    }

    const pureVersion = currentVersion.split('-')[0]
    const [major, minor, patch] = pureVersion.split('.').map(Number)
    return ({
        major: () => `${major + 1}.0.0`,
        minor: () => `${major}.${minor + 1}.0`,
        patch: () => `${major}.${minor}.${patch + 1}`,
    })[releaseType]()
}
