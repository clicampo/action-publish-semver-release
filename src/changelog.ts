import { getInput } from '@actions/core'
import { getExecOutput } from '@actions/exec'
import { getOctokit } from '@actions/github'
import type { Context } from '@actions/github/lib/context'
import type { ReleaseType } from './version'
import { getReleaseTypeFromCommitMessage } from './version'

type CommitsByReleaseType = Record<ReleaseType, { message: string; url: string; author: string }[]>
type CommitList = Awaited<
ReturnType<
ReturnType<typeof getOctokit>['rest']['repos']['listCommits']
>
>['data']

const run = async(command: string) => (await getExecOutput(command)).stdout

const getLastCommits = async(context: Context) => {
    const githubToken = getInput('github-token') || process.env.GH_TOKEN
    if (githubToken === '' || githubToken === undefined)
        throw new Error('GitHub token is required')

    const github = getOctokit(githubToken).rest

    // get the sha of the last tagged commit
    const lastTag = await run('git describe --tags --abbrev=0')
    const lastTaggedCommitSha = await run(`git rev-list -n 1 ${lastTag}`)

    const { data: commits } = await github.repos.listCommits({
        owner: context.repo.owner,
        repo: context.repo.repo,
    })

    const lastCommits = []
    for (const commit of commits) {
        if (commit.sha === lastTaggedCommitSha)
            break
        lastCommits.push(commit)
    }

    return lastCommits
}

const groupCommitsByReleaseType = (commits: CommitList) => {
    return commits
        .map((commit) => {
            const { message, url, author } = commit.commit
            const type = getReleaseTypeFromCommitMessage(message)
            return { message, type, url, author: String(author?.name) }
        })
        .reduce((commitsByType, commit) => {
            if (commit.type === null)
                return commitsByType
            if (!(commit.type in commitsByType))
                commitsByType[commit.type] = []
            commitsByType[commit.type].push(commit)
            return commitsByType
        }, {} as (CommitsByReleaseType))
}

const formatCommitsByType = (commitsByType: CommitsByReleaseType) => {
    let changelog = ''
    const getCommitInfo = (commit: { message: string; url: string; author: string }) => {
        const message = commit.message.split(':')[1].trim()
        const scope = commit.message.match(/^(.*?): /)?.[1] ?? ''
        const commitSha = commit.url.split('/').pop()?.slice(0, 8)
        return { message, scope, commitSha }
    }
    if (commitsByType.major) {
        changelog += `${[
            '## ⚠️ This release introduces breaking changes',
            '### Features',
        ].join('\n')}\n`
    }
    if (commitsByType.minor) {
        if (!commitsByType.major)
            changelog += '### Features\n'

        const featureCommits = [
            ...(commitsByType.major || []),
            ...(commitsByType.minor || []),
        ]
        for (const commit of featureCommits) {
            const { message, scope, commitSha } = getCommitInfo(commit)
            changelog += `- **(${scope})** ${message} ([${commitSha}](${commit.url}))\n`
        }
    }
    if (commitsByType.patch) {
        changelog += '### Bug Fixes\n'
        for (const commit of commitsByType.patch) {
            const { message, scope, commitSha } = getCommitInfo(commit)
            changelog += `- **(${scope})** ${message} ([${commitSha}](${commit.url}))\n`
        }
    }

    return changelog
}

export const generateChangelog = async(context: Context) => {
    const lastCommits = await getLastCommits(context)
    const commitsByType = groupCommitsByReleaseType(lastCommits)
    return formatCommitsByType(commitsByType)
}
