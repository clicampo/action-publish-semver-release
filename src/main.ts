import * as core from '@actions/core'
import { context } from '@actions/github'
import { generateChangelog } from './changelog'
import { notifyDiscordChannel } from './discord'
import { getLastCommitMessage, getLastGitTag, tagCommit } from './git'
import { createGithubRelease } from './github'
import { notifySlackChannel } from './slack'
import { getNextVersion, getPureVersion, getReleaseTypeFromCommitMessage } from './version'

async function run(): Promise<void> {
    const isReleaseCandidate = core.getInput('release-candidate') === 'true'
    const slackWebhookUrl = core.getInput('slack-webhook-url')
    const discordWebhookUrl = core.getInput('discord-webhook-url')

    try {
        const currentVersion = await getLastGitTag({
            considerReleaseCandidates: true,
            logInGroup: true,
        })
        if (currentVersion === null)
            return

        const lastCommitMessage = await getLastCommitMessage()
        if (lastCommitMessage === null)
            return

        const releaseType = getReleaseTypeFromCommitMessage(lastCommitMessage)

        // If the commit isn't of type `feat` or `fix`, we don't want to bump the version
        if (releaseType !== null) {
            const nextVersion = isReleaseCandidate
                ? getNextVersion({ currentVersion, releaseType })
                : (
                    currentVersion.match(/rc$/)
                        ? getPureVersion(currentVersion)
                        : getNextVersion({ currentVersion, releaseType })
                )
            core.info(`Publishing a release candidate for version ${nextVersion}`)

            const changelog = await generateChangelog(context)

            await tagCommit(nextVersion, isReleaseCandidate)

            await createGithubRelease(context, nextVersion, changelog, isReleaseCandidate)

            if (slackWebhookUrl !== '') {
                await notifySlackChannel(slackWebhookUrl, {
                    projectName: context.repo.repo,
                    projectUrl: core.getInput('project-url'),
                    productionActionUrl: core.getInput('production-action-url'),
                    nextVersion,
                    changelog,
                    isReleaseCandidate,
                })
            }
            if (discordWebhookUrl !== '') {
                await notifyDiscordChannel(discordWebhookUrl, {
                    projectName: context.repo.repo,
                    projectUrl: core.getInput('project-url'),
                    productionActionUrl: core.getInput('production-action-url'),
                    nextVersion,
                    changelog,
                    isReleaseCandidate,
                })
            }

            core.setOutput('next-version', nextVersion)
            core.setOutput('release-type', releaseType)
        }
        else {
            core.info('âœ… No new releases to be published!')
        }
    }
    catch (error) {
        if (error instanceof Error)
            core.setFailed(error.message)
    }
}

run()
