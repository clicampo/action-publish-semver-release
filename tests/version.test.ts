import { describe, expect, it } from 'vitest'
import { getNextVersion, getPureVersion } from '../src/version'

describe('Version should', () => {
    it('infer next version correctly', () => {
        expect(getNextVersion({ currentVersion: '1.0.0', releaseType: 'patch' }))
            .toMatch('1.0.1')
        expect(getNextVersion({ currentVersion: '1.0.0', releaseType: 'minor' }))
            .toMatch('1.1.0')
        expect(getNextVersion({ currentVersion: '1.0.0', releaseType: 'major' }))
            .toMatch('2.0.0')
        expect(getNextVersion({ currentVersion: '1.0.0-rc', releaseType: 'major' }))
            .toMatch('2.0.0')
    })

    it('throw error if version is in the wrong format', () => {
        expect(() => getNextVersion({ currentVersion: 'foo', releaseType: 'patch' }))
            .toThrowErrorMatchingSnapshot('"Invalid current version: foo"')
    })

    it('get pure version correctly', () => {
        expect(getPureVersion('1.0.0-rc-beta')).toMatch('1.0.0')
    })
})
