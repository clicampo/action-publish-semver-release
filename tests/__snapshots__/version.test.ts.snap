// Vitest Snapshot v1

exports[`Version should > throw error if version is in the wrong format > "Invalid current version: foo" 1`] = `"Invalid current version: foo"`;
